lib/spack/spack/package.py

```diff
@@ -1420,6 +1524,43 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class Package(PackageBase):
+    phases = ['install', 'log']
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install', 'log']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
+        return list()
```

Comment:
Taat doesn't answer my question.  The spack setup command looks for a stage
named 'configure'.  If people can name stages whatever they like, then
spack setup will not work on them.  I think we need a standardized set of
phase names so commands like 'install', 'setup', etc. can do their thing.
Whether we implement those names with subclassing and a million stub
methods, or with dynamic introspection, is an implementation detail.  I
vote for the former because it is standard OO technique.
On Jul 9, 2016 3:24 AM, "Denis Davydov" <notifications@github.com> wrote:

In lib/spack/spack/package.py:

> +
> +
> +class AutotoolsPackage(PackageBase):
> +    phases = ['autoreconf', 'configure', 'build', 'install', 'log']
> +
> +    def autoreconf(self, spec, prefix):
> +        """Not needed usually, configure should be already there"""
> +        pass
> +
> +    @PackageBase.sanity_check('autoreconf')
> +    def is_configure_or_die(self):
> +        if not os.path.exists('configure'):
> +            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
> +
> +    def configure_args(self):
> +        return list()

How would spack setup work if packages have non standardized phase names?

my understanding is that in these rare cases, instead of deriving from
AutotoolsPackage or CMakePackage classes, this single package can derive
from the PackageBase class and define whatever phases it needs. But it this
happens, most likely you came across a new build system or package type and
therefore should rather introduce another class similar to AutotoolsPackage
or CMakePackage.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/LLNL/spack/pull/1186/files/fa2783b9dc8c13e675a9b6791b240cdca61e4307#r70163831>,
or mute the thread
<https://github.com/notifications/unsubscribe/AB1cd18fEljn4GScEYhzStKWDDWQ1AElks5qT0yrgaJpZM4JG8iX>
.